### PR TITLE
BM-5 Friend Requests, cleaned up UI

### PR DESCRIPTION
**Pull request by umair-khalid-brainx**

### CODE REVIEW

File: app/assets/stylesheets/main.scss [hunk 1]
```diff
@@ -53,3 +57,49 @@
...
+  border: 1px solid #ff96ff;
+  padding: 15px 20px;
+}
+
+.nav-bar{
+  display: flex;
+  flex-direction: column;
+  padding: 0;
+}
+.nav-button{
```
Comment (sobanakram) [inline app/assets/stylesheets/main.scss:89]: @umair-khalid-brainx 
Kindly try to format code before committing code.
You can use shortcut to format it.

In above code there should be space before `{`

File: app/controllers/friend_requests_controller.rb [hunk 2]
```diff
@@ -0,0 +1,36 @@
...
+  end
+
+  def update
+    @friend_request = FriendRequest.find(params[:id])
+    @friend_request.accept
+    redirect_to friend_requests_path
+  end
+
+  def destroy
+    @friend_request = FriendRequest.find(params[:id])
```
Comment (sobanakram) [inline app/controllers/friend_requests_controller.rb:21]: Use before action for finding friendrequest object.

File: app/controllers/friend_requests_controller.rb [hunk 3]
```diff
@@ -0,0 +1,36 @@
...
+  def create_friend_request
+    friend = User.find(params[:user_id])
+    @friend_request = current_user.friend_requests.new(friend: friend)
+
+    if @friend_request.save
+      render :index, status: :created, location: @friend_request
+    else
+      redirect_to controller: :users, action: :index, notice: 'Cant send friend request'
+    end
+  end
```
Comment (sobanakram) [inline app/controllers/friend_requests_controller.rb:35]: Why we are not using `create` method for creating Friend request?

File: app/models/friend_request.rb [hunk 4]
```diff
@@ -1,4 +1,27 @@
 class FriendRequest < ApplicationRecord
-  belongs_to :requestor, class_name: :User
-  belongs_to :receiver, class_name: :User
+  belongs_to :user
+  belongs_to :friend, class_name: 'User'
+
+  validates :user, presence: true
+  validates :friend, presence: true, uniqueness: { scope: :user }
```
Comment (sobanakram) [inline app/models/friend_request.rb:6]: Good

File: app/models/friend_request.rb [hunk 5]
```diff
@@ -1,4 +1,27 @@
...
+  validate :not_self
+  validate :not_friends
+  validate :not_pending
+
+  def accept
+    user.friends << friend
+    destroy
```
Comment (sobanakram) [inline app/models/friend_request.rb:13]: I think, we shouldn't destroy friendrequest object,
Better solution would be to have status in FriendRequest and make it `pending`, `accepted` , `rejected`,
So, that we will have history as well.

File: app/models/friendship.rb [hunk 6]
```diff
@@ -1,4 +1,26 @@
 class Friendship < ApplicationRecord
-  belongs_to :friend_a, class_name: :User
-  belongs_to: :friend_b, class_name: :User
+  after_create :create_inverse_relationship
+  after_destroy :destroy_inverse_relationship
```
Comment (sobanakram) [inline app/models/friendship.rb:3]: Better to place callbacks after validations.

File: app/models/friendship.rb [hunk 7]
```diff
@@ -1,4 +1,26 @@
...
+  belongs_to :friend, class_name: 'User'
+
+  validates :user, presence: true
+  validates :friend, presence: true, uniqueness: { scope: :user }
+  validate :not_self
+
+  private
+
+  def create_inverse_relationship
+    friend.friendships.create(friend: user)
```
Comment (sobanakram) [inline app/models/friendship.rb:15]: Can we handle this without creating 2 Friendship objects?

File: app/models/user.rb [hunk 8]
```diff
@@ -10,9 +12,13 @@ class User < ApplicationRecord
...
+  has_many :friend_requests, dependent: :destroy
+  has_many :pending_friends, through: :friend_requests, source: :friend
 
-  has_many :friendships, ->(user) { where("friend_a_id = ? OR friend_b_id = ?", user.id, user.id) }
+  has_many :friendships, dependent: :destroy
   has_many :friends, through: :friendships
+
+  def remove_friend(friend)
+    current_user.friends.destroy(friend)
+  end
```
Comment (sobanakram) [inline app/models/user.rb:23]: How come we have access to `current_user` in Model?

It is only available in Controllers or Views

File: app/views/friend_requests/index.html.erb [hunk 9]
```diff
@@ -0,0 +1,43 @@
+<div class="display-5 my-5 text-center">Friend Requests</div>
+
+<div class="h3 my-5 text-center">Incoming Requests</div>
+
+<div class="container my-5">
+  <% unless @incoming.nil? %>
+    <% @incoming.each do |request| %>
+      <div class="user-entry">
+        <% incoming_friend = request.user %>
```
Comment (sobanakram) [inline app/views/friend_requests/index.html.erb:9]: There is N+1 query issue here,
Which needs to be fixed

File: app/views/friend_requests/index.html.erb [hunk 10]
```diff
@@ -0,0 +1,43 @@
...
+  <% end %>
+</div>
+
+<div class="h3 my-5 text-center">Outgoing Requests</div>
+
+<div class="container my-5">
+  <% unless @outgoing.nil? %>
+    <% @outgoing.each do |request| %>
+      <div class="user-entry">
+        <% outgoing_friend = request.friend %>
```
Comment (sobanakram) [inline app/views/friend_requests/index.html.erb:30]: There is N+1 query issue here,
Which needs to be fixed

File: app/views/friends/destroy.html.erb [hunk 11]
```diff
@@ -0,0 +1,2 @@
+<h1>Friends#destroy</h1>
```
Comment (sobanakram) [inline app/views/friends/destroy.html.erb:1]: Do we need this view?

File: config/routes.rb [hunk 12]
```diff
@@ -1,9 +1,22 @@
 Rails.application.routes.draw do
+  get 'friends/index'
+  get 'friends/destroy'
+
+  get 'friends#new', to: 'friends#create'
```
Comment (sobanakram) [inline config/routes.rb:5]: Should avoid using Routes like this,
better to use `resources` syntax.

File: config/routes.rb [hunk 13]
```diff
@@ -1,9 +1,22 @@
...
   devise_for :users
-  resources :users, only: [:index, :show]
   devise_scope :user do
     get '/users/sign_out' => 'devise/sessions#destroy'
   end
 
+  resources :users, only: [:index, :show]
+
+  # resources :friend_requests
```
Comment (sobanakram) [inline config/routes.rb:13]: Do we need this commented code?

File: app/views/friend_requests/index.html.erb [hunk 14]
```diff
@@ -0,0 +1,43 @@
...
+    <% @incoming.each do |request| %>
+      <div class="user-entry">
+        <% incoming_friend = request.user %>
+
+        <div class="col-1 h4 my-3 text-center"><%= image_tag incoming_friend.profile_photo, width: "30%", height: "30%" %></div>
+        <div class="col-2 h4 my-3 text-center"><%= incoming_friend.name %></div>
+        <div class="col-3 h4 my-3 text-center"><%= incoming_friend.email %></div>
+        <div class="col-3 h4 my-3"><%= incoming_friend.created_at.strftime("%d/%m/%Y, %I:%M %p") %></div>
+        <div class="col-3 h4 my-3 text-center">
+          <%= link_to "Accept", root_path, class: "small-button d-inline" %>
```
Comment (sobanakram) [inline app/views/friend_requests/index.html.erb:16]: Accepting friend request is working?

**Review by sobanakram — CHANGES_REQUESTED** [review]

File: app/controllers/friend_requests_controller.rb (hunk 3 above)
Comment (umair-khalid-brainx) [reply app/controllers/friend_requests_controller.rb:35]: I tried to direct to create method in routes but it was not working, consulted with Umer bhai as well. So made custom actions which are working fine

File: app/views/friend_requests/index.html.erb (hunk 14 above)
Comment (umair-khalid-brainx) [reply app/views/friend_requests/index.html.erb:16]: was resolved in next branches

File: app/models/friend_request.rb (hunk 5 above)
Comment (umair-khalid-brainx) [reply app/models/friend_request.rb:13]: Will work on this issue in latest branch

File: app/models/friendship.rb (hunk 7 above)
Comment (umair-khalid-brainx) [reply app/models/friendship.rb:15]: Will work on this issue in latest branch

**Pull request merged**